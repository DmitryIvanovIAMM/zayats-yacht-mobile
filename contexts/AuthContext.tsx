import { LoginCredentials } from "@/components/Login/LoginForm";
import {
  api,
  createLoginBody,
  mapObjectToFormUrlEncoded
} from "@/helpers/API/api";
import { API_PATHS } from "@/helpers/API/apiPaths";
import { PATHS } from "@/helpers/paths";
import { RelativePathString, useRouter } from "expo-router";
import { createContext, ReactNode, useContext, useState } from "react";

export interface UserInfo {
  user: {
    email: string;
    id: string;
    image: string | null;
    name: string | null;
  };
  expires: string;
}

export interface AuthState {
  isAuthenticated: boolean;
  userInfo?: UserInfo;
  isValidating?: boolean;
  error?: string;
}
export const defaultAuthState: AuthState = {
  isAuthenticated: false,
  userInfo: undefined,
  isValidating: false,
  error: undefined
};

export interface AuthContextType {
  authState: AuthState;
  login: (credentials: LoginCredentials, redirectPath: string) => void;
  logout: () => void;
  setAuthState: React.Dispatch<React.SetStateAction<AuthState>>;
}

// https://medium.com/@siddhantshelake/expo-router-authentication-with-protected-routes-persistent-login-eed364e310cc
const AuthContext = createContext<AuthContextType | null>(null);

export const AuthProvider = ({ children }: { children: ReactNode }) => {
  const [authState, setAuthState] = useState<AuthState>(defaultAuthState);
  const router = useRouter();
  const login = async (
    credentials: LoginCredentials,
    redirectPath: string = PATHS.landing
  ) => {
    try {
      setAuthState((prev) => ({
        ...prev,
        isValidating: true,
        error: undefined
      }));
      // https://github.com/nextauthjs/next-auth/issues/1110
      const csrfTokenRequest = await (
        await api.get(API_PATHS.GET_CSRF_TOKEN, undefined, true)
      ).json();
      const csrfToken = csrfTokenRequest.csrfToken;

      const body = createLoginBody(credentials, csrfToken);

      const loginRequest = await api.post(
        API_PATHS.SIGN_IN_WITH_CREDENTIALS,
        body,
        {
          "Content-Type": "application/x-www-form-urlencoded"
        },
        false,
        true
      );

      if (loginRequest.status === 200) {
        const getUserInfoSuccess = await getUserInfo();
        if (getUserInfoSuccess && redirectPath) {
          router.push(redirectPath as RelativePathString);
        }
      } else {
        return setAuthState({
          ...defaultAuthState,
          error: "Invalid credentials"
        });
      }
    } catch (error) {
      setAuthState({
        ...defaultAuthState,
        error: "Login failed"
      });
    } finally {
      setAuthState((prev) => ({
        ...prev,
        isValidating: false
      }));
    }
  };

  const getUserInfo = async (): Promise<boolean> => {
    try {
      const userInfo: UserInfo = await api
        .get(API_PATHS.GET_USER_SESSION, undefined, true)
        .then((res) => res.json());
      console.log("User Info: ", userInfo);

      if (!userInfo || !userInfo.user) {
        setAuthState({
          ...defaultAuthState,
          error: "Invalid credentials"
        });
        return false;
      }

      setAuthState({
        isAuthenticated: true,
        userInfo,
        isValidating: false,
        error: undefined
      });
      return true;
    } catch (error) {
      setAuthState({
        ...defaultAuthState,
        error: "Failed to fetch user info"
      });
      return false;
    }
  };

  const logout = async () => {
    try {
      setAuthState((prev) => ({
        ...prev,
        isValidating: true,
        error: undefined
      }));

      const csrfTokenRequest = await (
        await api.get("auth/csrf", undefined, true)
      ).json();
      const csrfToken = csrfTokenRequest.csrfToken;

      const logoutRequest = await api.post(
        API_PATHS.SIGN_OUT,
        mapObjectToFormUrlEncoded({
          csrfToken,
          redirect: false,
          json: true
        }),
        { "Content-Type": "application/x-www-form-urlencoded" },
        false,
        true
      );

      if (logoutRequest.status !== 200) {
        throw new Error("Failed to log out");
      }

      // remove this code, under logout; it only for debugging
      // const getUserInfoSuccess = await getUserInfo();
      // console.log("User Info after logout: ", getUserInfoSuccess);

      return setAuthState({
        ...defaultAuthState,
        isAuthenticated: false,
        userInfo: undefined,
        error: undefined
      });
    } catch (error) {
      setAuthState({
        ...defaultAuthState,
        error: "Logout failed"
      });
    } finally {
      setAuthState((prev) => ({
        ...prev,
        isValidating: false
      }));
    }
  };
  return (
    <AuthContext.Provider value={{ authState, login, logout, setAuthState }}>
      {children}
    </AuthContext.Provider>
  );
};

export const useAuth = (): AuthContextType => {
  const context = useContext(AuthContext);
  if (!context) {
    throw new Error("useAuth must be used within an AuthProvider");
  }
  return context;
};
