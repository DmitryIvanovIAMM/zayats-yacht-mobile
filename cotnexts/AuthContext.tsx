import { LoginCredentials } from "@/components/Login/LoginForm";
import { api, createLoginBody, mapObjectToFormUrlEncoded } from "@/helpers/api";
import { PATHS } from "@/helpers/paths";
import { RelativePathString, useRouter } from "expo-router";
import { createContext, ReactNode, useContext, useState } from "react";

export interface UserInfo {
  user: {
    email: string;
    id: string;
    image: string | null;
    name: string | null;
  };
  expires: string;
}

export interface AuthState {
  isAuthenticated: boolean;
  userInfo?: UserInfo;
  isValidating?: boolean;
  error?: string;
}
export const defaultAuthState: AuthState = {
  isAuthenticated: false,
  userInfo: undefined,
  isValidating: false,
  error: undefined,
};

export interface AuthContextType {
  authState: AuthState;
  login: (credentials: LoginCredentials, redirectPath: string) => void;
  logout: () => void;
  setAuthState: React.Dispatch<React.SetStateAction<AuthState>>;
}

// https://medium.com/@siddhantshelake/expo-router-authentication-with-protected-routes-persistent-login-eed364e310cc
const AuthContext = createContext<AuthContextType | null>(null);

export const AuthProvider = ({ children }: { children: ReactNode }) => {
  const [authState, setAuthState] = useState<AuthState>(defaultAuthState);
  const router = useRouter();
  const login = async (
    credentials: LoginCredentials,
    redirectPath: string = PATHS.landing
  ) => {
    try {
      setAuthState((prev) => ({
        ...prev,
        isValidating: true,
        error: undefined,
      }));
      // https://github.com/nextauthjs/next-auth/issues/1110
      const csrfTokenRequest = await (await api.get("auth/csrf")).json();
      const csrfToken = csrfTokenRequest.csrfToken;

      const body = createLoginBody(credentials, csrfToken);

      const loginRequest = await api.post(
        "auth/callback/credentials?",
        body,
        {
          "Content-Type": "application/x-www-form-urlencoded",
        },
        false
      );

      if (loginRequest.status === 200) {
        const getUserInfoSuccess = await getUserInfo();
        if (getUserInfoSuccess && redirectPath) {
          router.push(redirectPath as RelativePathString);
        }
      } else {
        return setAuthState({
          ...defaultAuthState,
          error: "Invalid credentials",
        });
      }
    } catch (error) {
      setAuthState({
        ...defaultAuthState,
        error: "Login failed",
      });
    } finally {
      setAuthState((prev) => ({
        ...prev,
        isValidating: false,
      }));
    }
  };

  const getUserInfo = async (): Promise<boolean> => {
    try {
      const userInfo: UserInfo = await api
        .get("auth/session")
        .then((res) => res.json());

      if (!userInfo || !userInfo.user) {
        setAuthState({
          ...defaultAuthState,
          error: "Invalid credentials",
        });
        return false;
      }

      setAuthState({
        isAuthenticated: true,
        userInfo,
        isValidating: false,
        error: undefined,
      });
      return true;
    } catch (error) {
      setAuthState({
        ...defaultAuthState,
        error: "Failed to fetch user info",
      });
      return false;
    }
  };

  const logout = async () => {
    try {
      setAuthState((prev) => ({
        ...prev,
        isValidating: true,
        error: undefined,
      }));

      const csrfTokenRequest = await (await api.get("auth/csrf")).json();
      const csrfToken = csrfTokenRequest.csrfToken;

      const logoutRequest = await api.post(
        "auth/signout",
        mapObjectToFormUrlEncoded({
          csrfToken,
          redirect: false,
          json: true,
        }),
        { "Content-Type": "application/x-www-form-urlencoded" },
        false
      );

      if (logoutRequest.status !== 200) {
        throw new Error("Failed to log out");
      }

      // remove this code, under logout; it ony for debugging
      // const getUserInfoSuccess = await getUserInfo();
      // console.log("User Info after logout: ", getUserInfoSuccess);

      return setAuthState({
        ...defaultAuthState,
        isAuthenticated: false,
        userInfo: undefined,
        error: undefined,
      });
    } catch (error) {
      setAuthState({
        ...defaultAuthState,
        error: "Logout failed",
      });
    } finally {
      setAuthState((prev) => ({
        ...prev,
        isValidating: false,
      }));
    }
  };
  return (
    <AuthContext.Provider value={{ authState, login, logout, setAuthState }}>
      {children}
    </AuthContext.Provider>
  );
};

export const useAuth = (): AuthContextType => {
  const context = useContext(AuthContext);
  if (!context) {
    throw new Error("useAuth must be used within an AuthProvider");
  }
  return context;
};
