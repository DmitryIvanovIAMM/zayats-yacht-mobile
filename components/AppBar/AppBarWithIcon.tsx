import { secondary } from "@/constants/Colors";
import { useAuth } from "@/contexts/AuthContext";
import { RelativePathString, useRouter } from "expo-router";
import * as React from "react";
import {
  ActivityIndicator,
  Image,
  Platform,
  StyleSheet,
  Text,
  View
} from "react-native";
import { Appbar } from "react-native-paper";
import WebAppBarWithIcon from "./WebAppBarWithIcon";

export interface AppBarWithIconProps {
  toggleMenu: () => void;
}

const AppBarWithIcon = ({ toggleMenu }: AppBarWithIconProps) => {
  const router = useRouter();
  const { logout, authState } = useAuth();

  const handleLoginPress = () => {
    if (authState.isAuthenticated) {
      logout();
    } else {
      router.push("/login" as RelativePathString);
    }
  };

  // Use web-specific AppBar for web platform
  if (Platform.OS === "web") {
    return <WebAppBarWithIcon toggleMenu={toggleMenu} />;
  }

  return (
    <Appbar.Header style={styles.container}>
      <View style={styles.firstDiv}>
        <Appbar.Action
          icon="menu"
          onPress={toggleMenu}
          color={secondary.dark}
        />
      </View>
      <Appbar.Content
        title={
          <View style={styles.logoContainer}>
            <Image
              // source={require("@/assets/images/zayats-logo-transparent.png")} //ZAYATS_embedded
              // source={require("@/assets/images/ZAYATS_embedded.svg")} //ZAYATS_embedded
              source={require("@/assets/images/zayats-logo-white.png")} //ZAYATS_embedded
              style={styles.alliedIcon}
            />
            <View
              style={styles.logoTouchable}
              onTouchEnd={() => router.push("/" as RelativePathString)}
            />
          </View>
        }
        style={styles.secondDiv}
      />
      <View style={styles.thirdDiv}>
        {authState.isValidating ? (
          <ActivityIndicator
            testID="spinner"
            size="small"
            color={secondary.dark}
            style={styles.activityIndicator}
          />
        ) : authState.isAuthenticated ? (
          <View style={styles.appBarIconWitName}>
            <View style={styles.userNameWithIcon}>
              <Text style={styles.userName}>
                {authState.userInfo?.user?.name
                  ? (authState?.userInfo?.user?.name ?? "User")
                  : (authState?.userInfo?.user?.email ?? "User")}
              </Text>
              <Appbar.Action
                icon="logout"
                onPress={handleLoginPress}
                color={secondary.dark}
                style={styles.appBarIconAfterNName}
              />
            </View>
          </View>
        ) : (
          <Appbar.Action
            icon="login"
            onPress={handleLoginPress}
            color={secondary.dark}
            style={styles.singAppBarIcon}
          />
        )}
      </View>
    </Appbar.Header>
  );
};

export default AppBarWithIcon;

const styles = StyleSheet.create({
  container: {
    backgroundColor: "white",
    display: "flex",
    flexDirection: "row",
    justifyContent: "space-between",
    alignItems: "center",
    width: "100%",
    ...Platform.select({
      web: {
        position: "sticky" as any,
        top: 0,
        zIndex: 1000,
        boxShadow: "0 2px 6px rgba(0,0,0,0.08)" as any
      },
      default: {
        elevation: 4
      }
    })
  },
  firstDiv: {
    flex: 1,
    justifyContent: "flex-start"
  },
  secondDiv: {
    flex: 1,
    alignItems: "center",
    justifyContent: "center"
  },
  thirdDiv: {
    flexDirection: "row",
    flex: 1,
    alignItems: "center",
    justifyContent: "flex-end"
  },
  alliedIcon: {
    height: 55,
    width: 80,
    maxWidth: "100%",
    bottom: 0,
    alignSelf: "center"
  },
  userNameWithIcon: {
    width: "100%",
    maxWidth: "100%",
    display: "flex",
    flexDirection: "row",
    justifyContent: "flex-end"
  },
  userName: Platform.select({
    default: {
      color: secondary.dark,
      fontSize: 18,
      marginRight: 8,
      fontWeight: "bold",
      textAlign: "right",
      alignSelf: "center"
    },
    android: {
      color: secondary.dark,
      fontSize: 14,
      fontWeight: "bold",
      textWrap: "wrap",
      width: "100%",
      textAlign: "right",
      alignSelf: "center"
    }
  }),
  activityIndicator: {
    marginRight: 16
  },
  singAppBarIcon: {
    padding: 0
  },
  appBarIconWitName: {
    flexDirection: "row",
    alignItems: "center"
  },
  logoContainer: {
    position: "relative",
    alignItems: "center",
    justifyContent: "center"
  },
  logoTouchable: {
    position: "absolute",
    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
    zIndex: 1
  },
  appBarIconAfterNName: {
    marginLeft: 0,
    padding: 0,
    width: "auto"
  },
  iconButton: {
    padding: 8,
    borderRadius: 4,
    minWidth: 40,
    minHeight: 40,
    justifyContent: "center",
    alignItems: "center",
    backgroundColor: "transparent"
  }
});
