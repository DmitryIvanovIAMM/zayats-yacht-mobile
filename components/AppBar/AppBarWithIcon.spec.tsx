import { useAuth } from "@/contexts/AuthContext";
import { fireEvent, render } from "@testing-library/react-native";
import React from "react";
import AppBarWithIcon from "./AppBarWithIcon";

// Mock react-native-paper Appbar with simple primitives
jest.mock("react-native-paper", () => {
  const React = require("react");
  const { View, TouchableOpacity } = require("react-native");
  const Appbar: any = {};
  Appbar.Header = ({ children, ...p }: any) =>
    React.createElement(View, p, children);
  Appbar.Action = ({ icon, onPress, ...p }: any) =>
    React.createElement(TouchableOpacity, {
      accessibilityLabel: String(icon),
      onPress,
      ...p,
    });
  Appbar.Content = ({ title, ...p }: any) =>
    React.createElement(View, p, title);
  return { Appbar };
});

// Mock expo-router with an internal mockPush exposed via the module
jest.mock("expo-router", () => {
  const mockPush = jest.fn();
  return {
    __esModule: true,
    useRouter: () => ({ push: mockPush }),
    mockPush,
  };
});

// Mock AuthContext with switchable return
jest.mock("@/contexts/AuthContext", () => ({
  useAuth: jest.fn(),
}));

describe("AppBarWithIcon", () => {
  const toggleMenu = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
    // default: logged out, not validating
    (useAuth as jest.Mock).mockReturnValue({
      logout: jest.fn(),
      authState: {
        isAuthenticated: false,
        isValidating: false,
        userInfo: undefined,
      },
    });
  });

  it("matches snapshot (logged out)", () => {
    const { toJSON } = render(<AppBarWithIcon toggleMenu={toggleMenu} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it("triggers toggleMenu when menu icon is pressed", () => {
    const { getByLabelText } = render(
      <AppBarWithIcon toggleMenu={toggleMenu} />
    );
    fireEvent.press(getByLabelText("menu"));
    expect(toggleMenu).toHaveBeenCalled();
  });

  it("shows login icon when logged out and navigates to /login on press", () => {
    const { getByLabelText } = render(
      <AppBarWithIcon toggleMenu={toggleMenu} />
    );
    fireEvent.press(getByLabelText("login"));
    const { mockPush } = require("expo-router");
    expect(mockPush).toHaveBeenCalledWith("/login");
  });

  it("shows spinner when validating and hides login/logout", () => {
    (useAuth as jest.Mock).mockReturnValue({
      logout: jest.fn(),
      authState: {
        isAuthenticated: false,
        isValidating: true,
        userInfo: undefined,
      },
    });
    const { queryByLabelText, toJSON } = render(
      <AppBarWithIcon toggleMenu={toggleMenu} />
    );
    expect(queryByLabelText("login")).toBeNull();
    expect(queryByLabelText("logout")).toBeNull();
    expect(toJSON()).toMatchSnapshot();
  });

  it("shows user name and logout icon when authenticated, and logs out on press", () => {
    const logout = jest.fn();
    (useAuth as jest.Mock).mockReturnValue({
      logout,
      authState: {
        isAuthenticated: true,
        isValidating: false,
        userInfo: { user: { name: "John Doe", email: "john@example.com" } },
      },
    });

    const { getByText, getByLabelText } = render(
      <AppBarWithIcon toggleMenu={toggleMenu} />
    );

    expect(getByText("John Doe")).toBeTruthy();
    fireEvent.press(getByLabelText("logout"));
    expect(logout).toHaveBeenCalled();
  });

  it("falls back to email when name is not provided", () => {
    (useAuth as jest.Mock).mockReturnValue({
      logout: jest.fn(),
      authState: {
        isAuthenticated: true,
        isValidating: false,
        userInfo: { user: { email: "john@example.com" } },
      },
    });

    const { getByText } = render(<AppBarWithIcon toggleMenu={toggleMenu} />);
    expect(getByText("john@example.com")).toBeTruthy();
  });
});
