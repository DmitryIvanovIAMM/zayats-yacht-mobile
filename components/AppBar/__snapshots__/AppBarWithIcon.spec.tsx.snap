// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppBarWithIcon matches snapshot (logged out) 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "white",
      "display": "flex",
      "elevation": 4,
      "flexDirection": "row",
      "justifyContent": "space-between",
      "width": "100%",
    }
  }
>
  <View
    style={
      {
        "flex": 1,
        "justifyContent": "flex-start",
      }
    }
  >
    <View
      accessibilityLabel="menu"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    />
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <Image
      source={
        {
          "testUri": "../../../assets/images/zayats-logo-white.png",
        }
      }
      style={
        {
          "alignSelf": "center",
          "bottom": 0,
          "height": 55,
          "width": 80,
        }
      }
    />
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "flex-end",
      }
    }
  >
    <View
      accessibilityLabel="login"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
          "padding": 0,
        }
      }
    />
  </View>
</View>
`;

exports[`AppBarWithIcon shows spinner when validating 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "white",
      "display": "flex",
      "elevation": 4,
      "flexDirection": "row",
      "justifyContent": "space-between",
      "width": "100%",
    }
  }
>
  <View
    style={
      {
        "flex": 1,
        "justifyContent": "flex-start",
      }
    }
  >
    <View
      accessibilityLabel="menu"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    />
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <Image
      source={
        {
          "testUri": "../../../assets/images/zayats-logo-white.png",
        }
      }
      style={
        {
          "alignSelf": "center",
          "bottom": 0,
          "height": 55,
          "width": 80,
        }
      }
    />
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "flex-end",
      }
    }
  >
    <ActivityIndicator
      color="#006666"
      size="small"
      style={
        {
          "marginRight": 16,
        }
      }
    />
  </View>
</View>
`;
